src/probe.bpf.c

```diff
@@ -28,257 +28,520 @@ struct {
-  bpf_probe_read_user(&state->command, arg_copy_len, (char*) arg_start);
-  for (int i = 0; i < arg_copy_len; i++) {
-    if (state->command[i] == '\0') {
-      state->command[i] = ' ';
+  return true;
+}
+
+static inline bool process_dns_answers(
+    struct inflight_dns_query* state,
+    const struct dnshdr* header,
+    struct cursor* body)
+{
+  int addresses_recorded = 0;
+  for (int i = 0; i < bpf_ntohs(header->ans_count) && i < 10; i++) {
+    if (jump_dns_name(body)) {
+      bpf_printk("[dnssnoop] Failed to jump domain name.");
+      return false;
+    }
+
+    // Type
+    u16* type_ptr = bpf_dynptr_slice(body->buf, body->offset, NULL, 2);
+    if (type_ptr == NULL) {
+      return false;
+    }
+    body->offset += 2;
+    u8 type = bpf_ntohs(*type_ptr);
+
+    // Class
+    u16* class_ptr = bpf_dynptr_slice(body->buf, body->offset, NULL, 2);
+    if (class_ptr == NULL) {
+      return false;
     }
+    u16 class = bpf_ntohs(*class_ptr);
+    body->offset += 2;
+
+    // TTL, just skip for now.
+    body->offset += 4;
+
+    // RDLength
+    u16* rd_len_ptr = bpf_dynptr_slice(body->buf, body->offset, NULL, 2);
+    if (rd_len_ptr == NULL) {
+      return false;
+    }
+    body->offset += 2;
+    u16 rd_len = bpf_ntohs(*rd_len_ptr);
+
+    // RData, based on type.
+    if (type == 0x01) {
+      if (rd_len != 4) {
+        return false;
+      }
+      u32* rd_ptr = bpf_dynptr_slice(body->buf, body->offset, NULL, 4);
+      if (rd_ptr == NULL) {
+        return false;
+      }
+      if (addresses_recorded < RESOLVED_ADDRESSES_MAX) {
+        BPF_SNPRINTF(state->resolved_addresses[addresses_recorded++], IP_BUF_SIZE, "%pI4", rd_ptr);
+      }
+    }
+    body->offset += rd_len;
   }
 
-  if (arg_len > arg_copy_len) {
-    state->command[COMMAND_BUF_SIZE - 1] = '\0';
-    state->command[COMMAND_BUF_SIZE - 2] = '>';
-    state->command[COMMAND_BUF_SIZE - 3] = '.';
-    state->command[COMMAND_BUF_SIZE - 4] = '.';
-    state->command[COMMAND_BUF_SIZE - 5] = '.';
-    state->command[COMMAND_BUF_SIZE - 6] = '<';
+  return true;
+}
+
+static bool process_dns_authority(
+    struct inflight_dns_query* state,
+    const struct dnshdr* header,
+    struct cursor* body)
+{
+  return false;
+}
+
+static bool process_dns_additional(
+    struct inflight_dns_query* state,
+    const struct dnshdr* header,
+    struct cursor* body)
+{
+  return false;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Process DNS Body
+////////////////////////////////////////////////////////////////////////////////
+
+static bool process_dns_body(
+    struct inflight_dns_query* state,
+    struct dnshdr* header,
+    struct bpf_dynptr* body_ptr)
+{
+  struct cursor body = { body_ptr, 0 };
+  if (!process_dns_questions(state, header, &body)) {
+    return false;
   }
+  if (!process_dns_answers(state, header, &body)) {
+    return false;
+  }
+
+  // Not yet implemented:
```

Comment:
just take out?

Reply:
🤷 , I think we'll get there eventually and having this in there will just help with context in the future. LMK if you feel strongly that it should go.